### ISSUE TITLE
Disassembly of vbroadcastsd memory access is incorrect

### ISSUE THREAD

**Issue by TheBlupper** [Vector35/binaryninja-api#4617]
**Version and Platform (required):**
- Binary Ninja Version: 3.4.4514-dev
- OS: Ubuntu Linux
- OS Version: 23.04
- CPU Architecture: x64

**Bug Description:**
The disassembly of `vbroadcastsd`'s source memory address is sometimes incorrect. See the [attached binary](https://github.com/Vector35/binaryninja-api/files/12545659/sahuang.gz) at `0x17d7`. Binary Ninja disassebmles the instruction to `vbroadcastsd zmm0, k0, qword [rsp-0xa]`, while all other disassemblers I've tried disassembles the same instruction to `vbroadcastsd zmm0,QWORD PTR [rsp-0x50]`.

This may very well be the case for many other similarly encoded instructions, but I have not tested those myself.

**Steps To Reproduce:**
 - Open the attached binary
 - Go to 0x17d7
 - View the disassembly

**Expected Behavior:**
Binary Ninja disassembles it correctly.

**Comment by plafosse** [comment]
Likely due a bug in xed. We plan on updating xed in this release cycle.

**Comment by xusheng6** [comment]
I think our disassembly is actually correct. Looking at the bytes,

```
000017d7  62f2fd48194424f6   vbroadcastsd zmm0, k0, qword [rsp-0xa]
```

It starts with byte 0x62, and according to https://en.wikipedia.org/wiki/EVEX_prefix, this is an `EVEX` prefix (as opposed to the `VEX` prefix, which starts 0xc4 or 0xc5). And according to https://www.felixcloutier.com/x86/vbroadcast, the EVEX version of vbroadcastsd is defined as:


```
EVEX.512.66.0F38.W1 19 /r VBROADCASTSD zmm1 {k1}{z}, xmm2/m64 	B 	V/V 	AVX512F
```

So, technically, there is nothing wrong about our disassembly. 

A subtle issue here is the register `k0` is special, and when it is used in such instructions, it always means no masking. So the disassembly output from other tools are also correct. In other words, `k0` is similar to the zero register in MIPS. One might ask, then why wouldn't we always hide `k0`?

This can be explained by the following paragraph from the [AVX-512](https://en.wikipedia.org/wiki/AVX-512#Opmask_registers) page:

```
Most AVX-512 instructions may indicate one of 8 opmask registers (k0–k7). For instructions which use a mask register as an opmask, register 'k0' is special: a hardcoded constant used to indicate unmasked operations. For other operations, such as those that write to an opmask register or perform arithmetic or logical operations, 'k0' is a functioning, valid register.
```

`k0` can be used like regular registers in instructions like `kmovb`, etc. For example, checkout https://travisdowns.github.io/blog/2019/12/05/kreg-facts.html.

This means if we wish to hide the `k0` when it is used as an mask, we would need to differentiate between the two classes of instructions, which I do not see an easy way to do.

**Comment by TheBlupper** [comment]
The issue in the disassembly produced is not that `k0` isn't omitted, it's that the source address is `[rsp-0xa]` instead of `[rsp-0x50]`

**Comment by xusheng6** [comment]
> The issue in the disassembly produced is not that `k0` isn't omitted, it's that the source address is `[rsp-0xa]` instead of `[rsp-0x50]`

Oh my bad, my eyes were caught by that `k0` and totally ignored the last operand. Re-opening the issue

**Comment by xusheng6** [comment]
xed itself seems to be decoding the instruction correctly:

```
$ obj/wkit/examples/obj/xed-ex1 -64 62f2fd48194424f6
Attempting to decode: 62 f2 fd 48 19 44 24 f6 
iclass VBROADCASTSD	category BROADCAST	ISA-extension AVX512EVEX	ISA-set AVX512F_512
instruction-length 8
operand-width 64
effective-operand-width 64
effective-address-width 64
stack-address-width 64
iform-enum-name VBROADCASTSD_ZMMf64_MASKmskw_MEMf64_AVX512
iform-enum-name-dispatch (zero based) 4
iclass-max-iform-dispatch 6
Nominal opcode position 4
Nominal opcode 0x19
Operands
#   TYPE               DETAILS        VIS  RW       OC2 BITS BYTES NELEM ELEMSZ   ELEMTYPE   REGCLASS
#   ====               =======        ===  ==       === ==== ===== ===== ======   ========   ========
0   REG0             REG0=ZMM0   EXPLICIT   W      ZF64  512    64     8     64     DOUBLE        ZMM
1   REG1               REG1=K0   EXPLICIT   R      MSKW   64     8    64      1        INT       MASK
2   MEM0           (see below)   EXPLICIT   R         Q   64     8     1     64     DOUBLE    INVALID
Memory Operands
  0    read BASE= RSP/GPR DISPLACEMENT_BYTES= 2 0xffffffffffffffb0 base10=-80 ASZ0=64
  MemopBytes = 8
ATTRIBUTES: DISP8_TUPLE1 MASKOP_EVEX MEMORY_FAULT_SUPPRESSION 
66-OSZ PREFIX
MANDATORY 66 PREFIX
EXCEPTION TYPE: AVX512_E6
BROADCAST
AVX512
Vector length: 512
AVX512 vector elements: 8
ISA SET: [AVX512F_512]
0	CPUID GROUP NAME: [AVX512F_512_AVX10]
	0	CPUID RECORD NAME: [AVX10_ENABLED]
		{Leaf 0x00000007, subleaf 0x00000001, EDX[19:19]} = 1
	1	CPUID RECORD NAME: [AVX10_VER1]
		{Leaf 0x00000024, subleaf 0x00000000, EBX[0:7]} = 1
	2	CPUID RECORD NAME: [AVX10_512VL]
		{Leaf 0x00000024, subleaf 0x00000000, EBX[18:18]} = 1
1	CPUID GROUP NAME: [AVX512F_512]
	0	CPUID RECORD NAME: [AVX512F]
		{Leaf 0x00000007, subleaf 0x00000000, EBX[16:16]} = 1
```

We can see `base10=-80` reports the correct offset

**Comment by xusheng6** [comment]
This is very tricky and I am still trying to fully understand it. Here is my progress so far:

1. This is not a bug in xed itself. I tried both the current version of xed that we are using, as well as the latest version, and if I check the output from the `xed-ex1` program, the correct value is returned
2. We are reporting a displacement of 0xa (10), and the correct displacement is 0x50 (80), they are different by a multiplier of 8. This is not a coincidence. For AVX512 instructions, the displacement should be multiplied by 8
3. However, we are already using the high-level API `xed_decoded_inst_get_memory_displacement` to retrieve the displacement. The differences in the AVX512 instructions should have been already accounted for. Also, the working xed CLI, i.e., `xed-ex1`, also uses the very same API and it gets 80 on that instruction. So we are NOT misusing the xed API. 
4. Now it seems we are building xed in a way that is subtly wrong. There are two versions of the same function, `xed_operand_values_get_memory_displacement_int64`, one being common https://github.com/intelxed/xed/blob/d41e876ba1bda5e53af2db74700216da29752fd7/src/common/xed-operand-values-interface-repl.c#L52-L58, and another one for AVX512 specifically: https://github.com/intelxed/xed/blob/a3055cd0209f5c63c88e280bbff9579b1e2942e2/datafiles/avx512f/xed-operand-values-interface-uisa.c#L68-L102. I tested that, when I run the xed CLI, the AVX512 verion is used, but if I run it within binja, the "common" one is used. So no wonder we get a value without the 8 multiplier. However, I am still trying to figure out what is causing this difference

**Comment by xusheng6** [comment]
OK finally found the cause: https://github.com/intelxed/xed/blob/a3055cd0209f5c63c88e280bbff9579b1e2942e2/datafiles/avx512f/files.cfg#L56-L58

The xed build system replaces one C source file with another during configuration. Applying the same replacement to the list of source code files seem to fix the problem

**Comment by xusheng6** [comment]
Should be fixed in dev 3.6.4785
